### ISSUE TITLE
[question] Take latest version of requirement fulfilling condition

### ISSUE THREAD

**Issue by iaavo** [conan-io/conan#11700]
I have two Conan packages: `lib` and `headers` in such a dependency graph:
```
 ______________________                 _______________
|                      |               |               |
|    lib/2022-07-25    | ------------> |  headers/1.0  |
|______________________|               |_______________|

 ______________________                        ^
|                      |                       |
|    lib/2022-07-26    | ----------------------/
|______________________|

 ______________________                 _______________
|                      |               |               |
|    lib/2022-07-27    | ------------> |  headers/2.0  |
|______________________|               |_______________|
```
Now I have a Conan package `app`. How can I make `app` require the latest version of `lib` which is still using `headers/1.0`? Note that I cannot control the version of `lib`.

**Comment by memsharded** [comment]
Hi @iaavo 

It is not fully clear what you mean. If you have ``app``, it will contain a ``requires = "lib/2022-07-25"`` (by the way, you should probably not be using ``-`` here, as it has some meaning in ``semver``, meaning pre-release, and it could affect how this version is interpreted in some cases). So it is the ``app`` that control the version of ``lib``.

**Comment by iaavo** [comment]
The use case that I have is that I want to support multiple maintenance versions of `app` at the same time. 
For example:

`app/1.0` is for anything that is built using `headers/1.0`
`app/2.0` is for anything that is built using `headers/2.0`

Now during development a new version of `lib` can be released. By checking the dependencies of the new package I can find out which `headers` the lib was based on and trigger the corresponding build. I.e. if the requirements of the incoming `lib` package contains a `headers/1.0`, I know that I need to trigger the `app/1.0` build.

When the `app/1.0` build is then run, I would like Conan to use the latest available version of `lib` which is still based on `headers/1.0` without having to adapt the recipe of `app`.

**Comment by memsharded** [comment]
Sorry, it is still not clear.
even if you have 2 versions of app, you still have ``requires`` in them. Lets say you have:

*app/1.0*
```
class App(ConanFile):
      requires = "lib/version1"
```
*app/2.0*
```
class App(ConanFile):
     requires = "lib/version2"
```

So the version is still explicit.
If you want, you could have a version range like: ``requires = "lib/[>1.0 <2.0]"``, but for a start, you are using a versioning scheme that is not compatible with version ranges, so this wouldn't be a solution for your case, and you would need to use explicit versions.

Regarding on "what needs to be rebuilt", I suggest having a look to: https://docs.conan.io/en/latest/versioning/lockfiles/ci.html
The idea is that knowing the dependants, or descendants of a given package version is not only difficult to know (it requires storing the inverse graph model somewhere), but it is also intractable in the majority of cases. So the recommended approach is the "products" approach: you have a set of applications and versions that you want to maintain, lets say in your case, your products are ``app/1.0`` and ``app/2.0``. Then, for each of your products, you can do a ``conan create`` or ``conan install --build=missing``, to potentially integrate latest changes from new released version in the dependencies (assuming that there are compatible version ranges).

Otherwise, there is no explicit way to do: bring a "lib" dependency that is compatible with "headers" dependency. Dependency resolution algorithm has other mechanisms, like version conflicts and overriding (you can depend on ``headers``, the version that you want, directly from the ``app``, and different things can happen:
- In Conan 1.X it will override/force that version upstream. 
- In Conan 1.X you can force that to fail if you set the ``general.error_on_override`` configuration. 
- In Conan 2.0 it will automatically error if the version is incompatible, unless you explicitly define ``override=True``.

In short, in your case, I think the approach that I would go for is every ``app`` version explicitly defines which ``lib`` version it is using, as it is not possible to use version ranges anyway with your versions, this is necessary.

**Comment by iaavo** [comment]
Thanks for the quick and elaborate reply, really amazing support here! I was hoping there was some internal Conan mechanism to resolve such a dependency graph with somewhat arbitrary conditions, but seems there isn't (and probably for good reason).

After looking through the Conan manual and some more thoughts to this issue, I think what I will try to make is a pre-CI script which will analyse new `lib` packages and create an alias depending on its `header` requirement. Something like that:
```
                                                          ______________________          _______________
                                                         |                      |        |               |
                                                         |    lib/2022-07-25    | -----> |  headers/1.0  |
                                                         |______________________|        |_______________|
                                                          
 _______________          ______________________          ______________________                 ^
|               |        |                      |        |                      |                |
|    app/1.0    | -----> |      lib/(1.x)       | -----> |    lib/2022-07-26    | ---------------/
|_______________|        |______________________|        |______________________|
                                                          
 _______________          ______________________          ______________________          _______________
|               |        |                      |        |                      |        |               |
|    app/2.0    | -----> |      lib/(2.x)       | -----> |    lib/2022-07-27    | -----> |  headers/2.0  |
|_______________|        |______________________|        |______________________|        |_______________|
```
This would mean additional effort to properly managing the aliases and always keeping them up to date but I think this approach could work for me.

As such, this issue can be closed, unless you have some doubts on this approach.

**Comment by memsharded** [comment]
> Thanks for the quick and elaborate reply, really amazing support here! I was hoping there was some internal Conan mechanism to resolve such a dependency graph with somewhat arbitrary conditions, but seems there isn't (and probably for good reason).

Thanks, we really love to help the best we can, sometimes it is challenging because of quick growth, but we always try :)

Yes, indeed. Arbitrary conditions means a way more complicated algorithm, actually could be NP-complete, and require joint-compatibility and backtracking, making it easily intractable in practice, as very often resolving dependencies means hitting a server, downloading a conanfile.py and its files, unzipping them, loading it with the Python interpreter, etc.


While the approach you outline seems could be doable, I am not sure about the tradeoff, the necessary effort and maintenance vs the value.  Of course it depends on many things, but probably you are not creating a major ``app/X.0`` version everyday? If not, what is really the effort to define the ``lib`` explicitly?
If the concern is having to maintain 2 different versions of the ``app`` ``conanfile.py`` recipe, why is it? It is very normal to have the app conanfile.py evolve with the repo itself, the code also changes, and there are many other things in the recipe that might need change between major versions. And if it doesn't, it is also possible to parameterize the recipe with a ``conandata.yml`` (we use this approach in ConanCenter, for example).

I am talking from the outside and without enough knowledge, of course, sure there are many things that I am missing. But a general recommendation of starting "lean" and "manual", but simple, is something that we have learned it is good for our users.

Sure, lets close this issue, but don't hesitate to re-open or create a new issue to further discuss things. Thanks!
